emerleite/tsuru-bluegreen
# Workaround tsuru's issue when removing multiple units

# Description

There's an issue when performing the swap of an app with multiple units. After the first unit is removed, the following error is encountered when trying to remove the remaining processes:

`event locked: app(user-tests-green) running "app.update.unit.remove" start by user yago.user@u.user at 2020-08-17T18:19:44Z`

According to tsuru's team, this is probably due to the project's lock scheme.

More about the changes can be found in the README.

# Changes Summary

- Add a number of retries to the `remove_units_per_process_type` method;
- Make it configurable (and optional) through the `.ini` file;

# How to test

Unit tests were added, so:
```
$ make testdeps  && make test
```

or, if you wish to use Docker:
```
$ make docker-test
```

One can also try swapping a test-app to see the timeouts and retries (don't forget to customize your `.ini` file as mentioned in this project's README).
